### DISCUSSION TITLE
Rethinking the local library approach

### DISCUSSION THREAD

**Discussion by NattyNarwhal** [SubmarinerApp/Submariner#201] [Polls]
Since from the days before I developed Submariner, it's had features for managing a local library, in addition to music from a server. While this aggregates downloaded music from a server, you can also import music from the local filesystem, be it copied into the directory or linked to the original file. I'm not sure why it was like that, but it is how it was.

However, I've been thinking of getting rid of this. It complicates a lot of things with the interface and code that makes it harder to work on. Specifically:

* Linked files mean I have to poke holes in the sandbox by keeping bookmarks (GH-197), and present a good UI for this. We don't do this, so it's broken outside of the user's music directory. The fact this is broken until I do this or drop the feature is what's currently blocking the next release.
* Importing local files means we have to read metadata within the file, and currently we have an ugly multi-strategy bit of to deal with this. With the problems this approach has, we'll probably have to import TagLib eventually.
* There's some UI inconsistencies with server vs. local music (i.e. how indices work, none of the other views) that would be hard to resolve in a satisfactory way. They kinda still exist if importing local music is removed, but the de-emphasis makes it a lower priority.
* I don't use the feature much myself, so it falls on the wayside with testing.
* More philosophically, It's a Subsonic server focused app, and keeping it lightweight and focused is my ideal.

Removing it would allow Submariner to focus on managing music from Subsonic servers, and allow removing a lot of code. If I do such a thing:

* The local library continues to exist for providing a way to view and manage downloaded tracks.
* Local files that do exist wouldn't be changed, there just wouldn't be a way to add more. This keeps the schema identical.
* I might provide a way to clean up orphaned tracks, similar to what exists for cleaning up local tracks associated with a track that was downloaded.
  * Right now, tracks that are associated with a track that was downloaded are orphaned in the local library. This might be changed to be cleaned up on server deletion perhaps.
* Local playlists would exist to provide a way to have a playlist that has tracks between multiple servers.
* I would recommend applications like [Cog](https://cog.losno.co) or [Doppler](https://brushedtype.co/doppler/) for people who want to manage non-server music, or just have all your music managed by your server.

(As a lighter alternative, removing linked tracks might be another alternative that would keep things simpler.)

I however, want to hear from users before I do a bigger change like this. Such a change would probably entail a 4.0, and if I do go through with it, I might soft-deprecate the options (i.e. hiding behind a defaults database option) before totally removing it.

**Comment by raxelbrof** [comment]
So I'm assuming this is related to the bug I noted a couple of months back where local (SMB Volume) files wouldn't play after Submariner had been quit and relaunched. It would be a shame to ditch non-Subsonic, directly referenced files for a few reasons:

- i haven't seen any other media player that has problems with files like this, eg Swinsian (though it hasn't had a release in a while).
- Is Subsonic still a live project? Everything I've heard suggests it's effectively abandonware.
- I've never used Plex, Subsonic, Jellyfin or whatever - they all seemed too intimidating for a single computer where I don't want the audio on a NAS or network server.
- Submariner is the nicest looking audio player I've seen,  and really looks like it has the bones to succeed iTunes.
- Would making your own Subsonic fork / Submariner Library app be an option, so we could get your quality of art for the server app (that isn't tumbleweeds and ghost town), with a Mac native UI, as well as the player?
- For a reference to the above, there's a CBR / CBZ comic book reader for Desktop / iOS called YACReader, which has both a reading app, and a library app YACReader Library which can serve on-desktop, and to iOS. Something to consider?

**Reply by NattyNarwhal** [reply]
Sorry for the late reply, I took a break and got busy and answering this fell to the wayside.

It's a bit unusual you're using it just for local music. Submariner was AFAIK always intended to work with a server, and the local management part was an odd curiosity that's kinda second-class. The issue with the sandboxing did make me think why it's even there in the first place, prompting this post (I had also hid linking files until I fixed the issue or removed support, to avoid exposing broken functionality). If you're only using it for a local library, I think you should either look into setting up a Subsonic (or compatible) server, or use an app that focuses on local library management specifically (i.e. Doppler).

WRT Subsonic maintenance: I've written to Sindre before, so presumably it's maintained. However, there's an ecosystem of Subsonic compatible servers. I personally use Navidrome instead, because it's lightweight and avoids a lot of file organization issues.

**Reply by raxelbrof** [reply]
No worries :)

The thing about music servers is, I'd be happy to use one - and Subsonic looked like it was a great option for having Podcasts and Audiobooks, I just don't want to have to run an additional piece of hardware (I'm trying to consolidate ono a single workstation) to do it, and I've had trouble trying to find out what Subsonic actually looks like for a mac user - there's lots of "set up your Docker container..." type guides, etc.

I went a long way down the almost purchasing a NAS route, before I realised I just didn't need storage and serving running when my Mac wasn't awake already.

I've looked around at a number of server options and there seem to be a lot of compromises - I don't want to manage my library in a web browser, or with an Electron app - that's just not a vision of computing, or media management I'm interested in. Hence why I liked Submariner so much - it looked & felt like a proper *Mac* app.

> I personally use Navidrome instead, because it's lightweight and avoids a lot of file organization issues.

OK their documentation is really nice - alright, it's a commandline app, and you just... put music in whatever directory structure you like into its music directory, and it makes it available, then the client app uses the id3 tags to present it? Is that the sort of thing it is?

I'm certain I don't know enough about using media servers, but my experience so far has been a lot of really basic stuff is assumed knowledge.

> or use an app that focuses on local library management specifically (i.e. Doppler).

OH MY GOD, THANKYOU!

You have no idea how much time I've spent trying to find a Mac music player / manager that can direct sync to an iPhone, without having to jank the media files via iCloud, or through the Files app as an intermediate stage - and given the way search has been destroyed in the last couple of years. I'd never even heard of Doppler, and I've been researching this stuff weekly for at least a year. Wow. I'll check it out...

...then try to find an audiobook solution that doesn't require m4b files...
...then a podcast solution that can sync direct from computer to iPhone, AND integrate my existing downloaded files from shows with new downloads.

cheers.
